DateCalculation/src/main/java/com/example/form/ReferenceDateForm.java

```diff
@@ -1,13 +1,15 @@
 package com.example.form;
 
 import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.Pattern;
 
 import lombok.Data;
 
 @Data
 public class ReferenceDateForm {
 
 	@NotBlank
+	@Pattern(regexp = "([1-9][0-9]{3})(0[1-9]|1[0-2])(0[1-9]|[12][0-9]|3[01])")
```

Comment:
![badge](https://img.shields.io/badge/review-ask-blue.svg)
`@DateTimeFormat`使ってみるのはどうですか？

Reply:
うまくいかなかったでしょうか？

Reply:
以下のようにしないと`a20200101b`のパターンでもマッチしちゃいますね。
```suggestion
	@Pattern(regexp = "^([1-9][0-9]{3})(0[1-9]|1[0-2])(0[1-9]|[12][0-9]|3[01])$")
```

ウェブで正規表現をチェックするツールがあったりしますのでご活用ください。
https://regex101.com/r/uCHwFp/1
`regex checker online`と検索するといろいろでます。
それから、上記ツールを使う際、`Javaの正規表現`であることを確認してくださいね。
正規表現にも仕様がありものによって微妙に違うんですね。

Reply:
情報過多かな・・・ 😓

Reply:
@DataTimeFormatですが、最初に@NotBlankを使用していた関係で上手くいかなかったため、諦めて@Patternを使用しておりましたが、@NotNullに変えたところエラーが起こらずバリデーション&バインドが上手くいきました。

Reply:
ただ、`@DataTimeFormat`の導入に伴い、`checkReferenceDate`メソッドからリダイレクト先の`executeCalc`メソッドへのDate型の受け渡しができず、リダイレクト前にDate型⇒String型への変換を行った関係で、少しコードが冗長になってしまったかなという感じもします。

Reply:
ここのリダイレクト時のデータの受け渡しについてかなり時間をかけて調べ、`@DataTimeFormat`と`@PathVariable`を使用した記事をもとにトライしてみたのですが、結果的にデータの受け渡しでエラーが発生してしまっておりました。
リダイレクト先にDate型や任意の型の要素の受け渡しは可能なのでしょうか？
もし可能であれば参考記事等を共有していただけるとありがたいです。

Reply:
"([1-9][0-9]{3})(0[1-9]|1[0-2])(0[1-9]|[12][0-9]|3[01])"
正規表現についてはそれぞれ文字列の先頭と末尾を表す「^」と「$」が抜けているため、結果として指定した文字列が含まれていれば前後にどのような文字が入っていようが問題なしと判断される状態になっていたことが理解できました。
ただ、コードを修正せずにアプリ内で試しに「a20220207b」のような形でフォームに入力をした際でも、なぜかバリデーションエラーが表示されましたが、これはspringの他の機能が作用したものということでしょうか。

Reply:
> リダイレクト先にDate型や任意の型の要素の受け渡しは可能なのでしょうか？

PathVariableではないのですが、こちらだと不都合でしょうか？
https://spring.pleiades.io/spring-framework/docs/current/javadoc-api/org/springframework/web/servlet/mvc/support/RedirectAttributes.html
https://www.baeldung.com/spring-redirect-and-forward

Reply:
> 前後にどのような文字が入っていようが問題なしと判断される状態になっていた

こちら私も初めて知ったのですが問題ない動きでした。
結論として`^`、`$`はなくて大丈夫です。

どうして？という話ですが、
Javaで正規表現にマッチしているかどうかを判断するメソッドにMathcers#matchesがあります。
https://docs.oracle.com/javase/jp/11/docs/api/java.base/java/util/regex/Matcher.html#matches()
ドキュメントに
> 領域全体をこのパターンとマッチします。

と書いていますので「領域全体」がマッチしているかどうか判断するんですね。

それから@Pattherアノテーションが利用しているバリデーターはhibernateのPatternValidatorというものを使っているようなのですが、
該当のクラスのisValidメソッドを見るとMatcher#matchesメソッドを利用していることが確認できました。

<img width="844" alt="スクリーンショット 2022-02-08 13 33 09" src="https://user-images.githubusercontent.com/62045457/152918726-c65354c9-6b77-4579-a0c8-6a1a1ca2f9d4.png">

Reply:
>ただこれだとデータの受け渡しでエラーになるんですよね？
>LocalDateで試してもダメそうか試してほしいです。
こちらについては`return "redirect:/calculation/result/" + new SimpleDateFormat("yyyyMMdd").format(form.getReferenceDate());`
このコードでDate型のままでもエラーは出ませんでした。
リダイレクト先での受け取り方の問題で、Date型のものをDate型のまま受け取ろうとしていたためエラーが出ていたと解釈しています。
`@PathVariable`ではStirngやint等の比較的単純（すごく初学者的な表現ですが）な型でしか受け取ることができないという解釈でよろしかったでしょうか？

Reply:
4e7a223083485bc24c376cc7dbb423a3deb70484
こちらで修正させていただきました。
改めてDate型について調べてみると、公式でも非推奨になっていたり、LocalDate型が便利であることが分かりました。
またRedirectAttributesの便利さには大変驚きました。
二点修正したことでコードもかなりすっきりした気がします。

Reply:
https://github.com/ka-kamada/DateCalculation/blob/4e7a223083485bc24c376cc7dbb423a3deb70484/DateCalculation/src/main/java/com/example/service/impl/DateCalculationServiceImpl.java#L32-L40
ここの部分はだいぶ無理やり作ったなという感じがしており、APIの仕様に関するページはあまり見慣れていないのでまだ解読できていないのですが、`TemporalAmount`を使用して無駄なコードをなくしていけるような気がしているのですが、見当違いではなかったでしょうか？

Reply:
正規表現については今の時点ではなかなか理解できるレベルに達していないので、引き続き継続的に学習していきます。

Reply:
> TemporalAmountを使用して無駄なコードをなくしていけるような気がしているのですが、見当違いではなかったでしょうか？


いくつか改善したい点がある＆テストがしっかりできていないですが例えばこんな感じでしょうか。
```java
DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy年MM月dd日");
for (Formula formula : list) {
  String result = referenceDate.plus(Period.of(formula.getYear(), formula.getMonth(), formula.getDay())).format(dtf);
  formula.setDateCaluculationResult(result);
}
```

Reply:
> @PathVariableではStirngやint等の比較的単純（すごく初学者的な表現ですが）な型でしか受け取ることができないという解釈でよろしかったでしょうか？

いえ。LocalDateでもDateTimeFormatでpatternを指定すれば引数に受け取れますよ。

Reply:
13054c33afaea7c3a955c7e763653ee8618fd454
こちら修正しました！（小山さんのコード丸写しになってしまいましたが）
自分なりにAPIリファレンスを見て考えてみたものの、これ以上のものは思いつきませんでした。
拡張for文とメソッドチェーンを使ってここまでコードが整うものかと驚いております。
どちらも必須の書き方ではないために、使いどころを見落としてしまいそうですが、覚えるために今後は積極的に使っていこうと思います。

Reply:
>いえ。LocalDateでもDateTimeFormatでpatternを指定すれば引数に受け取れますよ。

すいません。質問が間違えておりました。
正しくは「redirectでデータを渡す際はStirngやint等の比較的単純な型でしか受け渡すことができないという解釈でよろしかったでしょうか？」です。
String型以外の参照型については、文字として出力すると複雑な文字列になってしまうことが多いので、その型のままパスに乗せるとredirect先で受け取ることができないのかなと思いまして。
データ型をそのままでredirect先に受け渡すのであればRedirectAttributesを使う必要があるという認識です。

Reply:
> 拡張for文とメソッドチェーンを使ってここまでコードが整うものかと驚いております。

恐縮です。ありがとうございます。
やろうと思えば返り値を工夫すればラムダ式を使って1行にもできますね。
それはまた次の機会にご紹介しますのでまたレビュー依頼してください！
改善できそうなところはご提案します！

Reply:
> 「redirectでデータを渡す際はStirngやint等の比較的単純な型でしか受け渡すことができないという解釈でよろしかったでしょうか？」

意味がわかりました。

> String型以外の参照型については、文字として出力すると複雑な文字列になってしまうことが多いので、その型のままパスに乗せるとredirect先で受け取ることができないのかなと思いまして。

そのとおりですね。パス変数は文字列で渡さないといけませんね。
私が言いたかったのは例えば以下のように受け取る先のメソッドでLocalDate型に変換した形で受け取れるはずですという意味でした。

※下記コードはイメージです。動作確認していないのでコピペでは動かないかも
```
@PostMapping("/form")
public String postForm() {
  LocalDate now = LocalDate.now();
  return "redirect:/result/" + now.toString();
}

@GetMapping("/result/{date}")
public String showResult(@DateTimeFormat("yyyyMMdd") @PathVariable("date") LocalDate date) {
   return ...
}
```

Reply:
https://github.com/ka-kamada/DateCalculation/blob/f4826d2d57c35e9d580e42599058e8ff5dfb3ca1/DateCalculation/src/main/java/com/example/controller/DateCalculationController.java#L36-L56

RedirectAttributes導入に伴い上書きされていますが、パス変数の受け取りについてはこの部分でできていたかと思います！

>そのとおりですね。パス変数は文字列で渡さないといけませんね。

ありがとうございます。
当初この部分の認識ができていなかったせいで、Date型のままredirectしたりしていて大変苦しみました。
「この感じは無理そうだな」と思いつつも、クリティカルなネット記事が見つからずだいぶ沼にハマっていたので、検索能力についても経験値がまだ足りていないように思います。
RedirectAttributesを教えていただいたおかげで、Controller間での値の受け渡しについては、かなり知識が整理されました！

Reply:
>それはまた次の機会にご紹介しますのでまたレビュー依頼してください！
>改善できそうなところはご提案します！

ありがとうございます！
引き続きこちらのアプリをもとに機能追加やテストコードのレビュー依頼も出させていただくことになると思います。
メソッドチェーンはかなり意識的に使っていこうと思いますが、「ここで使えるのに使えていない」「むしろここで使うと可読性が下がる」というような部分があればご指摘いただければと。

Reply:
もろもろ状況理解しました！
検索能力も経験あるのみです！
プログラミングは死にゲーなのでどんどん死にましょう笑